audition/thresholding.py

```diff
@@ -0,0 +1,86 @@
+import logging
+
+
+def _above_min(df, metric_filter):
+    return df[df['raw_value'] >= metric_filter['min_value']]
+
+
+def _close_to_best(df, metric_filter):
+    return df[df['below_best'] < metric_filter['max_below_best']]
+
+
+def _of_metric(df, metric_filter):
+    return df[
+        (df['metric'] == metric_filter['metric']) &
+        (df['parameter'] == metric_filter['metric_param'])
+    ]
+
+
+class ModelGroupThresholder(object):
+    def __init__(
+        self,
+        db_engine,
+        distance_from_best_table,
+        train_end_times,
+        initial_model_group_ids,
+    ):
+        self.db_engine = db_engine
+        self.distance_from_best_table = distance_from_best_table
+        self.train_end_times = train_end_times
+        self._model_group_ids = initial_model_group_ids
+        self._metric_filters = []
+
+    def _filter_model_groups(self, df, filter_func):
+        passing = set(self._model_group_ids)
+        for metric_filter in self._metric_filters:
+            passing &= set(filter_func(
+                _of_metric(df, metric_filter),
+                metric_filter
+            )['model_group_id'])
+        return passing
+
+    def model_groups_above_min(self, df):
+        return self._filter_model_groups(df, _above_min)
+
+    def model_groups_close_to_best(self, df):
+        return self._filter_model_groups(df, _close_to_best)
+
+    def model_groups_passing_rules(self):
+        below_min_model_groups = set(self._model_group_ids)
```

Comment:
As I was writing this docstring I realized that this logic won't allow the user to make the rules less restrictive; the more restricted rows will be in self._model_group_ids, and since that is passed, instead of increasing the size of the list it will just stay the same.

Adding a test to make sure that doesn't happen.  To fix, I'll either save both the initial list in the object alongside the current list, or just avoid caching the *current* list of model groups; it seems to be of uncertain value. The `model_group_ids` property can just always compute it based on the current metric filters and the original list, and caching the current list might be such a trivial speedup it's not worth the bug risk.